### DISCUSSION TITLE
Disabling / enabling an input action trough code temporarily

### DISCUSSION THREAD

**Discussion by Swarkin** [godotengine/godot-proposals#5674] [Scripting]
It would really help me if there was a way to temporarily disable and enable an action trough code, without making clunky bools for many inputs to check in my code manually. Personally, this would make my code cleaner and easier to work with.
I've imagined something like `InputMap.disable_action(action)` and `InputMap.enable_action(action)`.

**Comment by Calinou** [comment]
Out of curiosity, why do you need to temporarily disable an input? If you want to prevent player movement during a certain game state, it's better rely on the current game state or player properties instead.

**Reply by Swarkin** [reply]
Then I'd have to make a new variable just for checking if I can make a certain action, which is what I wanted to avoid.
Your recommendation doesn't fit well into my current project..
I just need a quick and easy way of disabling an action for a sec in certain situations and this would be the best way for it.
Just disabling an input action directly is better than checking in multiple scripts if some value is true or not, if you understand what I mean..

**Reply by Calinou** [reply]
> I just need a quick and easy way of disabling an action for a sec in certain situations and this would be the best way for it.

The reason I'm asking for a *precise* use case is to avoid [XY Problem](https://xyproblem.info/) situations :slightly_smiling_face:

**Reply by Swarkin** [reply]
Ok I think I understand, basically I'm playing a short tween to zoom the camera and during that time I can't let certain actions take place, which is why instead of making a global variable to check if action x can happen and clutter my autoloads, I could just disable the input action for the entire game during this short time.
I can't really give more info because it's a prototype project so this quicker way of doing it would save me some time and keep it cleaner.
I agree with you that in bigger projects it might be better to rely on the game state tho.. just not in this scenario.

**Comment by dalexeev** [comment]
See [input processing order](https://docs.godotengine.org/en/stable/tutorials/inputs/inputevent.html). If the node that should block input is located correctly, then you can use the methods [Contol.accept_event](https://docs.godotengine.org/en/stable/classes/class_control.html#class-control-method-accept-event), [Viewport.set_input_as_handled](https://docs.godotengine.org/en/stable/classes/class_viewport.html#class-viewport-method-set-input-as-handled) or [SceneTree.set_input_as_handled](https://docs.godotengine.org/en/stable/classes/class_scenetree.html#class-scenetree-method-set-input-as-handled) to stop the further propagation of the input event. For example:

```gdscript
extends Camera2D

var _block_input = false

func _input(event):
    if _block_input: # You can also check which actions using is_action.
        get_viewport().set_input_as_handled()

func animate():
    _block_input = true
    # ...Your code here...
    _block_input = false
```

If you are absolutely sure that you need to block all input, then there is the [Viewport.gui_disable_input](https://docs.godotengine.org/en/stable/classes/class_viewport.html#class-viewport-property-gui-disable-input) property.

**Reply by Swarkin** [reply]
This is what I wanted to avoid..
> [...] "without making clunky bools for many inputs to check in my code manually."

**Comment by timothyqiu** [comment]
Add this script as an Autoload named `InputBlocker`, then you can use `InputBlocker.enable_action("name")` and `InputBlocker.disable_action("name")`.

```gdscript
extends Node

var disabled_actions := PoolStringArray()

func _input(event: InputEvent) -> void:
	for action in disabled_actions:
		if event.is_action(action):
			get_tree().set_input_as_handled()
			return


func disable_action(name: String) -> void:
	if not name in disabled_actions:
		disabled_actions.append(name)


func enable_action(name: String) -> void:
	var idx := disabled_actions.find(name)
	if idx != -1:
		disabled_actions.remove(idx)
```

**Reply by Swarkin** [reply]
That would work.. thanks

**Comment by AbdouMadjidi** [comment]
Hey, I wonder if this is a solution to stop focus toggling between buttons in a container when pressing tab. I'm trying to block the tab press from the buttons and their parent only and not from all the scene tree. None of the solutions here worked for me tho
Godot 4.4.1
